content/blog/2016/2016-04-07-pipeline-for-runs-on-hardware.adoc

```diff
@@ -0,0 +1,258 @@
+In this post I will focus on the Pipeline plugin.
+
+Below you can find an example of Pipeline scripts, which runs tests on FPGA board. The id of this board comes from build parameters (_fpgaId_). In this script we also presume that all nodes have pre-installed tools (Xilinx ISE in this case).
+
+```groovy
+// Run on node having my_fpga label 
+node("linux && ml509") {
+  git url:"http://github.com/oleg-nenashev/pipeline_hw_samples";
+  sh "make all";
+}
+```
+
+But such scenario could be also implemented in a Free-style project.
+What would we get from Pipeline plugin?
+
+## Getting added-value from Pipeline-as-code
+
+Pipeline provides much added-value features for hardware-based tests. 
+I would like to highlight the following advantages:
+
+* Robustness against restarts of Jenkins master.
+* Robustness against network disconnects. _sh()_ steps are based on the
+link:https://wiki.jenkins-ci.org/display/JENKINS/Durable+Task+Plugin[Durable Task plugin], so Jenkins can safely continue the execution flow once the node reconnects to the master.
+* It's possible to run tasks on multiple nodes w/o creating complex flows based on job triggers and copy artifact steps, etc. It can be achieved via combination of _parallel()_ and _node()_ stages.
+* Ability to store the shared logic in standalone Pipeline libraries
+* etc.
+
+First two advantages allow to improve the robustness of Jenkins nodes against infrastructure failures. 
+It is critical for long-running tests on hardware.
+
+Last two advantages address the flexibility of Pipeline flows.
+There are also plugins for freestyle projects, but they are not flexible enough.
+
+## Utilizing pipeline features
+
+The sample Pipeline script above is very simple. 
+We would like to get some added value from Jenkins.
+
+### General improvements
+
+Let's enhance the script by using several features being provided by pipeline in order to get visualization of stages, report publishing and build notifications.
+
+We also want to minimize the time being spent on the node with the attached FPGA board. 
+So we will split the bitfile generation and further runs to two different nodes in this case: a general purpose linux node, and the node with the hardware attached.
+
+You can find the resulting Pipeline script below:
+
+```groovy
+// Synthesize on any node
+def imageId="";
+node("linux") {
+  stage "Prepare environment" 
+  git url:"http://github.com/oleg-nenashev/pipeline_hw_samples";
+  // Construct the bitfile image ID from commit ID
+  sh 'git rev-parse HEAD > GIT_COMMIT'
+  imageId= "myprj-${fpgaId}-" + readFile('GIT_COMMIT').take(6)
+  
+  stage "Synthesize project"
+  sh "make FPGA_TYPE=$fpgaId synthesize_for_fpga"
+  /* We archive the bitfile before running the test, so it won't be lost it if something happens with the FPGA run stage. */
+  archive "target/image_${fpgaId}.bit"
+  stash includes: "target/image_${fpgaId}.bit", name: 'bitfile'
```

Comment:
You may find the syntax for "[Code block with callouts](http://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#source-code)" to be useful here. If you don't wish to use that syntax, I suggest adding code-comments for just about every line here to explain _why_ this Pipeline is doing each step so that those who are unfamiliar with Pipeline will more easily understand